pkg/image/admission/admission.go

```diff
@@ -0,0 +1,302 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+
+	kadmission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilerrors "k8s.io/kubernetes/pkg/util/errors"
+	"k8s.io/kubernetes/pkg/watch"
+	limitranger "k8s.io/kubernetes/plugin/pkg/admission/limitranger"
+
+	"github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	imagequota "github.com/openshift/origin/pkg/quota/image"
+)
+
+const (
+	PluginName = "ImageLimitRange"
+)
+
+func init() {
+	kadmission.RegisterPlugin(PluginName, func(client clientset.Interface, config io.Reader) (kadmission.Interface, error) {
+		plugin, err := NewImageLimitRangerPlugin(client, config)
+		if err != nil {
+			return nil, err
+		}
+		return plugin, nil
+	})
+}
+
+// imageLimitRangerPlugin is the admission plugin.
+type imageLimitRangerPlugin struct {
+	*kadmission.Handler
+	osClient    client.Interface
+	limitRanger kadmission.Interface
+
+	imageStreamReflector *cache.Reflector
+	imageStreamStore     cache.Store
+	stopChan             chan struct{}
+}
+
+// imageLimitRangerPlugin implements the LimitRangerActions interface.
+var _ limitranger.LimitRangerActions = &imageLimitRangerPlugin{}
+
+var _ = oadmission.WantsOpenshiftClient(&imageLimitRangerPlugin{})
+var _ = oadmission.Validator(&imageLimitRangerPlugin{})
+
+// NewImageLimitRangerPlugin provides a new imageLimitRangerPlugin.  Also see SetOpenshiftClient.
+func NewImageLimitRangerPlugin(client clientset.Interface, config io.Reader) (*imageLimitRangerPlugin, error) {
+	plugin := &imageLimitRangerPlugin{
+		Handler:          kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		imageStreamStore: cache.NewStore(cache.MetaNamespaceKeyFunc),
+	}
+	limitRanger, err := limitranger.NewLimitRanger(client, plugin)
+	if err != nil {
+		return nil, err
+	}
+	plugin.limitRanger = limitRanger
+
+	// NOTE: the reflector is set up in SetOpenshiftClient to ensure we have access to the client
+	return plugin, nil
+}
+
+// Run starts the reflectors used by the plugin.
+func (a *imageLimitRangerPlugin) Run() {
+	if a.stopChan == nil {
+		a.stopChan = make(chan struct{})
+		a.imageStreamReflector.RunUntil(a.stopChan)
+	}
+}
+
+// Stop stops the reflectors used by the plugin.
+func (a *imageLimitRangerPlugin) Stop() {
+	if a.stopChan != nil {
+		close(a.stopChan)
+		a.stopChan = nil
+	}
+}
+
+// Admit invokes the admission logic for checking against LimitRanges.
+func (a *imageLimitRangerPlugin) Admit(attr kadmission.Attributes) error {
+	if !a.SupportsAttributes(attr) {
+		return nil // not applicable
+	}
+
+	// If the namespace is not set on the object ensure it gets set.
```

Comment:
This feels all sorts of wrong - implies namespace should already be set.  I don't like admission controllers mutating this.